### PR TITLE
Introduce MvcRazorTemplateEngine

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <packageSources>
+  <add key="local" value="d:\work\katana\razor\artifacts\build" />
```
Comment (rynowak) [inline NuGet.config]: derp

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorCompilation.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.Mvc.Razor.Internal;
+using Microsoft.AspNetCore.Razor.Evolution;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Compilation
+{
+    public class MvcRazorCompilation : RazorCompilation
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorCompilation.cs]: The code that's in this class seems right, but it shouldn't be coupled to a specific item.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            get
+            {
+                if (_compilerCache == null)
+                {
+                    _compilerCache = _compilerCacheProvider.Cache;
+                }
+
+                return _compilerCache;
+            }
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs]: Why not just `=> _compilerCacheProvider.Cache`

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                message: error.GetMessage(),
+                formattedMessage: $"{error} ({sourceSpan.LineIndex},{sourceSpan.CharacterIndex}) {error.GetMessage()}",
+                filePath: filePath,
+                startLine: sourceSpan.LineIndex + 1,
+                startColumn: sourceSpan.CharacterIndex,
+                endLine: sourceSpan.LineIndex + 1,
+                endColumn: sourceSpan.CharacterIndex + sourceSpan.Length);
+        }
+
+        private static string ReadFileContentsSafely(RazorProjectItem projectItem)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs]: You don't need this if you already have a `RazorCodeDocument` or `RazorSourceDocument`.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCacheContext.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.Razor.Compilation;
+using Microsoft.AspNetCore.Razor.Evolution;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    public struct CompilerCacheContext
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCacheContext.cs]: Let's talk about this, this just seems a little like a turd.

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
*bump*

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs [hunk 6]
```diff
@@ -59,10 +50,23 @@ public RazorPageFactoryResult CreateFactory(string relativePath)
...
+            var result = _razorCompiler.Compile(relativePath);
             if (result.Success)
             {
-                return new RazorPageFactoryResult(result.PageFactory, result.ExpirationTokens, result.IsPrecompiled);
+                var compiledType = result.CompiledType;
+
+                var newExpression = Expression.New(compiledType);
+                var pathProperty = compiledType.GetTypeInfo().GetProperty(nameof(IRazorPage.Path));
+
+                // page.Path = relativePath;
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs]: Intentional?

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/CompilerCacheTest.cs [hunk 7]
```diff
@@ -54,12 +72,12 @@ public void GetOrAdd_ReturnsCompilationResultFromFactory()
...
 
             // Act
-            var result = cache.GetOrAdd(ViewPath, _ => expected);
+            var result = cache.GetOrAdd(ViewPath, CreateContextFactory(expected));
 
             // Assert
             Assert.True(result.Success);
-            Assert.IsType<TestView>(result.PageFactory());
-            Assert.Same(ViewPath, result.PageFactory().Path);
+            Assert.Equal(typeof(TestView), result.CompiledType);
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/CompilerCacheTest.cs]: I see it is just a not-null check. But just to put it out there, is it worth adding `Assert.True(result.Success)`?

Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/CompilerCacheTest.cs]: Isn't it doing that on the line above?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs]: Yup - it's sort of showing what the expression tree is generating. I can remove it if it's confusing or not obvious.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/CompilerCacheTest.cs (hunk 7 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/CompilerCacheTest.cs]: Yes it does! Github totally hid it from me 😕

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs (hunk 6 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs]: Got it. Maybe start with `Generates: ... ` or something similar.

**Pull request closed without merging**